### ISSUE TITLE
Set up continuous integration?

### ISSUE THREAD

**Issue by ctb** [datacarpentry/R-ecology-lesson#154]
While #151 wasn't caused by build errors, there were some Jekyll complaints that could have been caught by CI.  I'd be happy to put together a travis build for this repo if there is interest.

**Comment by fmichonneau** [comment]
yes, there is interest! I had set up a mock repo at https://github.com/fmichonneau/R-ecology-travis that is ready to go. From my recollection, we only need to create a "data carpentry bot user" that would be responsible for the commits and pushes of the builds (it seems better than using a real user account) and give it the appropriate write access privileges to the repo.

(duplicate of #23)

**Comment by tracykteal** [comment]
It would be great to set up TravisCI here, both so that we're making sure the current Rmd is the current md/HTML and to be able to test the build automatically before it's deployed. It could be a good model and trial for other lessons that are using RMarkdown too. @naupaka was just discussing this too.

I set up a 'datacarpentry-CI' github account and gave it Admin access to this repo. Email for ti goes to an alias. Let me know if you need the login info, or want to be added to the email alias.

**Comment by ctb** [comment]
Some thoughts, combined with some ideas bounced off of @tracykteal offline --

* without any changes to the current process, Travis could let alert us to
  any Jekyll warnings in PRs or on master.  This is pretty easy.

* to actually review content before merge, we'd have to find a place to stage
  built HTML that was separate from the github pages for this repo, b/c those
  can only display one branch.  I am not sure what all the options are here and
  will need to do some investigatin'.

* adding a split between branches and displayed content (as needed for reviewing
  before merge) could support having lesson versions.  @tracykteal also guessed
  that having a formal split between "in development" and "released" lessons
  might encourage people who are nervous about contributing to live lessons.

I'll poke around with the technical options here and see if I can come up with
a simple demo of "staging" vs "public" that supports rendering pull requests,
and then go from there.

**Comment by naupaka** [comment]
Linking this to [our discussion](https://github.com/swcarpentry/r-novice-gapminder/issues/222) about the same thing on the SWC r-novice-gapminder repo.

**Comment by tracykteal** [comment]
Also bringing in @butterflyology who is working on similar things for the Reproducible Research curriculum.

**Comment by tracykteal** [comment]
As @ctb said, I think having a split between 'development' and 'deployed' branches is a good idea, because it helps with quality assurance and I think could encourage contribution if there's less concern about things immediately going 'live'. I know that's a broader discussion than just on this issue, but it would be particularly important here, as Travis CI could go to 'dev' and then we could have a manual push to 'deployed'. 

The technical challenge as @ctb and others have identified is having a non gh-pages branch render as HTML somewhere so we can check formatting more effectively. I can imagine a few solutions, including having TravisCI go all the way to rendering the whole site (like a local Jekyll serve) that gets deployed somewhere, but maybe that's just a hack.

**Comment by ktoddbrown** [comment]
As someone looking to contribute without much background on git/jekyll, the CONTRIBUTING.md could use some work to make it clearer what the pipeline is. This was a bigger hurdle for me then any self consciousness about changes I suggested going to a live as opposed to a dev branch.

**Comment by tracykteal** [comment]
Thanks @ktoddbrown. That's really great and useful feedback. 

We also have been talking about clearer contributor guidelines, like [recommendations](https://opensource.com/life/16/3/contributor-guidelines-template-and-tips) from @captainsafia or [onboarding docs like rOpenSci](https://github.com/ropensci/onboarding). I'll start a separate issue on that and will link to it here.

**Comment by ctb** [comment]
See #159.

**Comment by ctb** [comment]
Spent some time on this today. Apologies for the TMI below, it gets technical :).

tl;dr? We really want something like readthedocs.io, but with support for Jekyll and PR branches. It's not possible currently.

In sum:

* to test pull requests, we'd have to write a .travis.yml that built PRs.

  A basic .travis.yml with a 'make' command in was already created by @fmichonneau so I just poached [his version](https://github.com/fmichonneau/R-ecology-travis/blob/gh-pages/.travis.yml) and removed stuff - [the result is here](https://github.com/ctb/R-ecology/pull/2). (Note, the builds are currently failing!)

  Next steps here would be to fix the build problems, and add Jekyll site builds to the R build.

  Then, merge in the .travis.yml change and then turn on travis on this repo.  After that, github settings could be adjusted so that merges could only be done after successful travis CI builds.

  This would serve the goal of checking PRs for basic correctness/syntax errors etc.

* to update the gh-pages branch automatically upon merge into (say) master, we'd have to fiddle with default branches and then adopt @fmichonneau's [full .travis.yml](https://github.com/fmichonneau/R-ecology-travis/blob/gh-pages/.travis.yml) for the master branch only.  It's reasonably complicated but might be a good thing; I'm agnostic on whether or not this worthwhile.

* to enable build & review of generated content on pull requests, we would have to be able to generate a static site in Travis, upload it somewhere, and report that location back to the pull request so that a reviewer could browse through it before merging.  This is currently not easy or robust enough for me to do ;).

   This is the path I pursued the furthest, so here are a few details on what went wrong:

  -  it's straightforward to host static Jekyll sites on Amazon S3, with automatic ~30 day turnover. Some useful bits are to use `jekyll build -b /some_pr_name` so that you can put the site in a subdirectory, and use `aws s3 sync _site s3://datacarpentry-ctb-test/some_pr_name -storage-class REDUCED_REDUNDANCY --acl public-read` to put the jekyll generated content on s3.

    - it is possible to use the `travis encrypt` functionality to privately store AWS keys etc. necessary for `aws s3 sync` to work. This allows PRs from within the repo, but . However, it is [currently impossible](https://github.com/travis-ci/travis-ci/issues/5579) to use `travis encrypt` for non-privileged user PRs, which makes this somewhat pointless for new users.

* an alternative would be to show people how to fork the repo into their own github repo and set up GitHub Pages there, but this has several limitations. One is that users could only have one rendered pull request at a time, because GitHub Pages can only be enabled on one branch per repo. Another is that setting up travis to update user repos with the build results is prohibitive in its technical complexity and probably not all that good an idea anyway.

* another alternative would be to set up a generic travis config that dumped each PR's generated site into something like [IPFS](https://ipfs.io/) where it could be viewed as a static site by anyone. In theory dumping static sites into IPFS is straightforward to do, but the tools to do this easily do not yet exist. We would still face the problem of reporting back the full site URL to the pull request but I think this is solvable with a bit of scripting. This is actually the direction I think is most useful to pursue in the medium term.  And with only a little bit of awesomeness it could include fully versioned data carpentry materials :).

I think that's it. I'll update if I remember something I missed.

**Comment by ctb** [comment]
I thought this was a neat diagram that is a workflow we might be able to get to work in some fashion:

https://github.com/harshjv/travis-ci-latex-pdf

![image](https://cloud.githubusercontent.com/assets/51016/21566556/da1e3338-ce58-11e6-8b30-0b001de41f3d.png)

**Comment by ctb** [comment]
Minor update for @fmichonneau and @tracykteal especially -- I had mentioned that readthedocs supports automatic building of documentation, plus versioning (e.g. [khmer v2.0 docs](https://khmer.readthedocs.io/en/v2.0/), [khmer v1.0 docs](https://khmer.readthedocs.io/en/v1.0/), [khmer master](https://khmer.readthedocs.io/en/latest/) which would give people stable long-term URLs as well as "look at the latest".  It's well integrated with github. It doesn't seem to support building off of pull requests but maybe we could ask them to implement that.  I should say that it's also a pretty stable setup and runs many of the Python world's docs.

The only problem is that it uses [sphinx](http://www.sphinx-doc.org/en/stable/) which is based on reStructuredText and doesn't support markdown. Or, rather, *didn't* - a few years ago I discovered that markdown was supported via a plug-in.

The [DIBSI site](http://ivory.idyll.org/dibsi/) is built using this plug-in.  It's entirely built by Sphinx, with a single .rst file to make Sphinx happy - see [the source](https://github.com/dib-lab/dibsi/). It may be a medium-term solution to some of the feature requests that @tracykteal has been talking about with me.  I'm happy to push on this if there's enthusiasm, or discuss blockers, or ...just let it sit here until next time I'm motivated :)

**Comment by tracykteal** [comment]
@jduckles pointed out that the SWC website is already set up this way, with the SWC website being deployed to software-carpentry.org from http://github.com/swcarpentry/website. So the hosting and TravisCI could be adapted for posting lessons. There's a new issue on this https://github.com/swcarpentry/lesson-example/issues/126

**Comment by fmichonneau** [comment]
the rendering (Rmd -> html) of the lessons is now done through travis-ci. The master branch (default) hosts the Rmd files, while the gh-pages branch hosts the rendered html.

I'll soon setup another branch to be able to render the development version of the lesson somewhere else, while the stable version will be where it currently lives. I'll close this issue then; and write up about this new setup.

**Comment by fmichonneau** [comment]
mostly implemented for now.
